Add pyhf inspect

kratsg opened the pull request:
# Description

An idea to inspect a given JSON and provide some summary information. Currently, output is as follows:

```
(pyhf) Lord Stark:~/pyhf (feature/inspect)$ pyhf inspect ~/Downloads/Excl-fit/Sig_MGPy8EG_A14N23LO_SM_Higgsino_130_100_2L2MET75_MadSpin_pruned.json 
                                                  Summary Statistics  
                                                  ------------------  
                                                     channels  22
                                                      samples  7
                                                   parameters  88
                                                    modifiers  122
                                                          poi  mu_SIG

                                                     channels  nbins
                                                   ----------  -----
                                          CRtau_hghmet_met_Et    1  
                                       SRee_eMLLf_hghmet_cuts    1  
                            SRee_eMLLd_lowmet_deltaM_low_cuts    1  
                           SRee_eMLLg_lowmet_deltaM_high_cuts    1  
                                           CRVV_hghmet_met_Et    1  
                            SRee_eMLLf_lowmet_deltaM_low_cuts    1  
                           SRee_eMLLc_lowmet_deltaM_high_cuts    1  
                            SRee_eMLLc_lowmet_deltaM_low_cuts    1  
                           SRee_eMLLh_lowmet_deltaM_high_cuts    1  
                                          CRtau_lowmet_met_Et    1  
                                       SRee_eMLLg_hghmet_cuts    1  
                                           CRVV_lowmet_met_Et    1  
                                       SRee_eMLLe_hghmet_cuts    1  
                                       SRee_eMLLh_hghmet_cuts    1  
                           SRee_eMLLe_lowmet_deltaM_high_cuts    1  
                            SRee_eMLLe_lowmet_deltaM_low_cuts    1  
                                          CRtop_hghmet_met_Et    1  
                           SRee_eMLLd_lowmet_deltaM_high_cuts    1  
                                          CRtop_lowmet_met_Et    1  
                                       SRee_eMLLd_hghmet_cuts    1  
                           SRee_eMLLf_lowmet_deltaM_high_cuts    1  
                                       SRee_eMLLc_hghmet_cuts    1  

                                                      samples
                                                   ----------
                                                        fakes
                                                      diboson
        MGPy8EG_A14N23LO_SM_Higgsino_130_100_2L2MET75_MadSpin
                                                        Zjets
                                                        other
                                                      Zttjets
                                                          top

                                                   parameters  constraint              modifiers
                                                   ----------  ----------              ----------
                                                       EG_Eff  constrained_by_normal   histosys,normsys
                                                       EG_Iso  constrained_by_normal   histosys,normsys
                                            EG_RESOLUTION_ALL  constrained_by_normal   histosys,normsys
                                                      EG_Reco  constrained_by_normal   histosys,normsys
                                                 EG_SCALE_AF2  constrained_by_normal   histosys,normsys
                                                 EG_SCALE_ALL  constrained_by_normal   histosys,normsys
                    JET_EtaIntercalibration_NonClosure_negEta  constrained_by_normal   histosys,normsys
                    JET_EtaIntercalibration_NonClosure_posEta  constrained_by_normal   histosys,normsys
                                          JET_Flavor_Response  constrained_by_normal   histosys,normsys
                                              JET_GroupedNP_1  constrained_by_normal   histosys,normsys
                                              JET_GroupedNP_2  constrained_by_normal   histosys,normsys
                                              JET_GroupedNP_3  constrained_by_normal   histosys,normsys
                                             JET_JER_DataVsMC  constrained_by_normal   histosys,normsys
                                        JET_JER_EffectiveNP_1  constrained_by_normal   histosys,normsys
                                        JET_JER_EffectiveNP_2  constrained_by_normal   histosys,normsys
                                        JET_JER_EffectiveNP_3  constrained_by_normal   histosys,normsys
                                        JET_JER_EffectiveNP_4  constrained_by_normal   histosys,normsys
                                        JET_JER_EffectiveNP_5  constrained_by_normal   histosys,normsys
                                        JET_JER_EffectiveNP_6  constrained_by_normal   histosys,normsys
                                JET_JER_EffectiveNP_7restTerm  constrained_by_normal   histosys,normsys
                                                          JVT  constrained_by_normal   normsys
                                         MET_SoftTrk_ResoPara  constrained_by_normal   histosys,normsys
                                         MET_SoftTrk_ResoPerp  constrained_by_normal   histosys,normsys
                                            MET_SoftTrk_Scale  constrained_by_normal   histosys,normsys
                                                 MUON_Eff_sys  constrained_by_normal   histosys,normsys
                                           MUON_Eff_sys_lowpt  constrained_by_normal   histosys,normsys
                                                      MUON_ID  constrained_by_normal   histosys,normsys
                                                      MUON_MS  constrained_by_normal   histosys,normsys
                                                   MUON_SCALE  constrained_by_normal   histosys,normsys
                                                   Norm_other  constrained_by_normal   normsys
                                               SigModelAndPDF  constrained_by_normal   normsys
                                                      SigXSec  constrained_by_normal   histosys,normsys
                                                      btag_BT  constrained_by_normal   histosys,normsys
                                                      btag_CT  constrained_by_normal   histosys,normsys
                                                   btag_Extra  constrained_by_normal   histosys,normsys
                                                  btag_LightT  constrained_by_normal   histosys,normsys
                                              fakes_flatSys20  constrained_by_normal   normsys
                                                         lumi  constrained_by_normal   lumi
                                                       mu_SIG  unconstrained           normfactor
                                                 mu_VV_hghmet  unconstrained           normfactor
                                                 mu_VV_lowmet  unconstrained           normfactor
                                                mu_Ztt_hghmet  unconstrained           normfactor
                                                mu_Ztt_lowmet  unconstrained           normfactor
                                                mu_top_hghmet  unconstrained           normfactor
                                                mu_top_lowmet  unconstrained           normfactor
                                                       pileup  constrained_by_normal   histosys,normsys
                shape_fakes_stat_fakes_CRVV_hghmet_obs_met_Et  constrained_by_poisson  shapesys
                shape_fakes_stat_fakes_CRVV_lowmet_obs_met_Et  constrained_by_poisson  shapesys
               shape_fakes_stat_fakes_CRtau_hghmet_obs_met_Et  constrained_by_poisson  shapesys
               shape_fakes_stat_fakes_CRtau_lowmet_obs_met_Et  constrained_by_poisson  shapesys
               shape_fakes_stat_fakes_CRtop_hghmet_obs_met_Et  constrained_by_poisson  shapesys
               shape_fakes_stat_fakes_CRtop_lowmet_obs_met_Et  constrained_by_poisson  shapesys
shape_fakes_stat_fakes_SRee_eMLLc_lowmet_deltaM_high_obs_cuts  constrained_by_poisson  shapesys
            shape_fakes_stat_fakes_SRee_eMLLd_hghmet_obs_cuts  constrained_by_poisson  shapesys
shape_fakes_stat_fakes_SRee_eMLLd_lowmet_deltaM_high_obs_cuts  constrained_by_poisson  shapesys
 shape_fakes_stat_fakes_SRee_eMLLd_lowmet_deltaM_low_obs_cuts  constrained_by_poisson  shapesys
            shape_fakes_stat_fakes_SRee_eMLLe_hghmet_obs_cuts  constrained_by_poisson  shapesys
shape_fakes_stat_fakes_SRee_eMLLe_lowmet_deltaM_high_obs_cuts  constrained_by_poisson  shapesys
 shape_fakes_stat_fakes_SRee_eMLLe_lowmet_deltaM_low_obs_cuts  constrained_by_poisson  shapesys
            shape_fakes_stat_fakes_SRee_eMLLf_hghmet_obs_cuts  constrained_by_poisson  shapesys
shape_fakes_stat_fakes_SRee_eMLLf_lowmet_deltaM_high_obs_cuts  constrained_by_poisson  shapesys
 shape_fakes_stat_fakes_SRee_eMLLf_lowmet_deltaM_low_obs_cuts  constrained_by_poisson  shapesys
            shape_fakes_stat_fakes_SRee_eMLLg_hghmet_obs_cuts  constrained_by_poisson  shapesys
shape_fakes_stat_fakes_SRee_eMLLg_lowmet_deltaM_high_obs_cuts  constrained_by_poisson  shapesys
            shape_fakes_stat_fakes_SRee_eMLLh_hghmet_obs_cuts  constrained_by_poisson  shapesys
shape_fakes_stat_fakes_SRee_eMLLh_lowmet_deltaM_high_obs_cuts  constrained_by_poisson  shapesys
                                 staterror_CRVV_hghmet_met_Et  constrained_by_normal   staterror
                                 staterror_CRVV_lowmet_met_Et  constrained_by_normal   staterror
                                staterror_CRtau_hghmet_met_Et  constrained_by_normal   staterror
                                staterror_CRtau_lowmet_met_Et  constrained_by_normal   staterror
                                staterror_CRtop_hghmet_met_Et  constrained_by_normal   staterror
                                staterror_CRtop_lowmet_met_Et  constrained_by_normal   staterror
                             staterror_SRee_eMLLc_hghmet_cuts  constrained_by_normal   staterror
                 staterror_SRee_eMLLc_lowmet_deltaM_high_cuts  constrained_by_normal   staterror
                  staterror_SRee_eMLLc_lowmet_deltaM_low_cuts  constrained_by_normal   staterror
                             staterror_SRee_eMLLd_hghmet_cuts  constrained_by_normal   staterror
                 staterror_SRee_eMLLd_lowmet_deltaM_high_cuts  constrained_by_normal   staterror
                  staterror_SRee_eMLLd_lowmet_deltaM_low_cuts  constrained_by_normal   staterror
                             staterror_SRee_eMLLe_hghmet_cuts  constrained_by_normal   staterror
                 staterror_SRee_eMLLe_lowmet_deltaM_high_cuts  constrained_by_normal   staterror
                  staterror_SRee_eMLLe_lowmet_deltaM_low_cuts  constrained_by_normal   staterror
                             staterror_SRee_eMLLf_hghmet_cuts  constrained_by_normal   staterror
                 staterror_SRee_eMLLf_lowmet_deltaM_high_cuts  constrained_by_normal   staterror
                  staterror_SRee_eMLLf_lowmet_deltaM_low_cuts  constrained_by_normal   staterror
                             staterror_SRee_eMLLg_hghmet_cuts  constrained_by_normal   staterror
                 staterror_SRee_eMLLg_lowmet_deltaM_high_cuts  constrained_by_normal   staterror
                             staterror_SRee_eMLLh_hghmet_cuts  constrained_by_normal   staterror
                 staterror_SRee_eMLLh_lowmet_deltaM_high_cuts  constrained_by_normal   staterror

```

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
- adds a CLI utility for inspecting information about the provided workspace object on an as-best basis
- testing for pyhf inspect is done by checking if the output is as expected for a provided xmlimport example
```

kratsg commented:
And if you provide `--output-file`, it'll jump the summary statistics in a JSON file for you as well, but I'm not really sure that's needed? Ideally, this was more for a human-friendly summary statistics.

matthewfeickert commented:
This will be quite helpful for debugging. We can ask that people submit it as part of their Issue.

lukasheinrich commented:
this looks good, but py27 seems to have some issues

kratsg commented:
@matthewfeickert , @lukasheinrich - this is ready to go. Here's a screenshot of what it looks like with the xmlimport validation example. I moved the `poi` information down into measurements and added an asterisk for the default measurement chosen.
  
<img width="1550" alt="Screenshot 2019-04-13 13 28 11" src="https://user-images.githubusercontent.com/761483/56079369-bebcef80-5df3-11e9-92d0-ca16b180988a.png">

lukasheinrich commented on pyhf/commandline.py:
```diff
@@ -100,6 +100,118 @@ def json2xml(workspace, output_dir, specroot, dataroot, resultprefix):
+
+    # summary statistics
+    fmtStr = '{{0: >{0:d}s}}  {{1:s}}'.format(maxlen + len('Summary'))
+    print(fmtStr.format('Summary Statistics', ''))
```
"Summary Statistics"  is a term of art in statistics, so it might be misleading here. I would just use `Summary` perhaps

lukasheinrich commented:
@kratsg I wonder if we should put this into its own module `inspect.py` that `commandline.py` imports such that it can be used as a library function

lukasheinrich commented on pyhf/commandline.py:
```diff
@@ -100,6 +100,115 @@ def json2xml(workspace, output_dir, specroot, dataroot, resultprefix):
+
+    # summary
+    fmtStr = '{{0: >{0:d}s}}  {{1:s}}'.format(maxlen + len('Summary'))
+    print(fmtStr.format('     Summary     ', ''))
```
perhaps instead of `print()` we should start using `click.echo()`

kratsg commented:
> @kratsg I wonder if we should put this into its own module `inspect.py` that `commandline.py` imports such that it can be used as a library function

Iffy on that. I justified that all the information used already exists in the API and there was just no CLI for it. It's not clear to me what a user needs from `inspect.py` that doesn't already exist on the `model.config` information (where almost everything except measurement name) comes from.

lukasheinrich commented:
I mean in the sense that workspace -> summary json is a transform you might want to do this in a non-CLI setting (imagine having to inspect an entire set of workspaces) .. I'm don't have a strong opinion on this (PR LGTM) but it's something to keep in mind

lukasheinrich commented on pyhf/pdf.py:
```diff
@@ -17,6 +17,7 @@ def __init__(self, spec, poiname='mu'):
         self.par_map = {}
         self.par_order = []
         self.next_index = 0
+        self.poi_name = None
```
why is this data member needed? I don't see it being used anywhere in the PR

kratsg replied:
This was added to make it easier to access the poi name instead of going through index -- mostly for other users. We do store poi_index, but it would be tricky for users to figure out how to get the poi name from that (and they don't necessarily need to know the internals of ModelConfig to do so).

kratsg commented:
> I mean in the sense that workspace -> summary json is a transform you might want to do this in a non-CLI setting (imagine having to inspect an entire set of workspaces) ..

I'll file an issue for this feature. It would be nice to provide a summary JSON that's useful for HEPData display -- e.g. we upload a workspace to HEPData, and they use the summary JSON feature to provide a quick summary of the likelihood JSON attachment.
